src/snowflake/cli/plugins/nativeapp/codegen/sandbox.py

```diff
@@ -0,0 +1,201 @@
+    return os.environ.get("CONDA_DEFAULT_ENV")
+
+
+def _is_venv_active() -> bool:
+    return _get_active_venv_dir() is not None
+
+
+def _is_conda_active() -> bool:
+    return _get_active_conda_env() is not None
+
+
+def _is_ms_windows() -> bool:
+    return sys.platform == "win32"
+
+
+def _execute_python_interpreter(
+    python_executable: Optional[Union[str, Path, Sequence[str]]],
+    script_source: str,
+    cwd: Optional[Union[str, Path]],
+    timeout: Optional[int],
+) -> subprocess.CompletedProcess:
+    if not python_executable:
+        raise SandboxExecutionError("No python executable found")
+
+    if isinstance(python_executable, str) or isinstance(python_executable, Path):
+        args = [python_executable]
+    else:
+        args = [arg for arg in python_executable]
+    args.append("-")
+    return subprocess.run(
+        args,
+        capture_output=True,
+        text=True,
+        input=script_source,
+        timeout=timeout,
+        cwd=cwd,
+    )
+
+
+def _execute_in_venv(
+    script_source: str,
+    venv_path: Optional[Union[str, Path]] = None,
+    cwd: Optional[Union[str, Path]] = None,
+    timeout: Optional[int] = None,
+) -> subprocess.CompletedProcess:
+    resolved_venv_path = None
+    if venv_path is None:
+        active_venv_dir = _get_active_venv_dir()
+        if active_venv_dir is not None:
+            resolved_venv_path = Path(active_venv_dir).resolve()
+    elif isinstance(venv_path, str):
+        resolved_venv_path = Path(venv_path).resolve()
+    else:
+        resolved_venv_path = venv_path.resolve()
+
+    if resolved_venv_path is None:
+        raise SandboxExecutionError("No venv root found")
+
+    if not resolved_venv_path.is_dir():
+        raise SandboxExecutionError(
+            f"venv path should be an existing directory: {resolved_venv_path}"
+        )
+
+    # find the python interpreter for this environment. There is no need to activate environment prior to invoking the
+    # interpreter, as venv maintains the invariant that invoking any of the scripts will set up the virtual environment
+    # correctly. activation scripts are only used for convenience in interactive shells.
+    if _is_ms_windows():
+        python_executable = resolved_venv_path / "Scripts" / "python.exe"
+    else:
+        python_executable = resolved_venv_path / "bin" / "python3"
+
+    if not python_executable.is_file():
+        raise SandboxExecutionError(
+            f"No venv python executable found: {resolved_venv_path}"
+        )
+
+    return _execute_python_interpreter(
+        python_executable, script_source, timeout=timeout, cwd=cwd
+    )
+
+
+def _execute_in_conda_env(
+    script_source: str,
+    env_name: Optional[str] = None,
+    cwd: Optional[Union[str, Path]] = None,
+    timeout: Optional[int] = None,
+) -> subprocess.CompletedProcess:
+    conda_env = env_name
+    if conda_env is None:
+        conda_env = _get_active_conda_env()
+    if conda_env is None:
```

Comment:
Nit: Nest under 112? Would prevent an additional check when `conda_env` is not None from the user.